controller/src/main/java/io/pravega/controller/store/stream/ZKStoreHelper.java

```diff
@@ -341,4 +372,54 @@ private BackgroundCallback callback(Consumer<CuratorEvent> result, Consumer<Thro
     PathChildrenCache getPathChildrenCache(String path, boolean cacheData) {
         return new PathChildrenCache(client, path, cacheData);
     }
+
+    <T> CompletableFuture<VersionedMetadata<T>> getCachedData(String path, String id, Function<byte[], T> fromBytes) {
+        return cache.getCachedData(new ZkCacheKey<>(path, id, fromBytes))
+                .thenApply(this::getVersionedMetadata);
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> VersionedMetadata<T> getVersionedMetadata(VersionedMetadata v) {
+        // Since cache is untyped and holds all types of deserialized objects, we typecast it to the requested object type
+        // based on the type in caller's supplied Deserialization function. 
+        return new VersionedMetadata<>((T) v.getObject(), v.getVersion());
+    }
+
+    void invalidateCache(String path, String id) {
+        cache.invalidateCache(new ZkCacheKey<>(path, id, x -> null));
+    }
+
+    @Data
```

Comment:
I think that normally annotations go immediately over the class declaration, right?